### PR TITLE
[M] 1514484: entitlements are regenerated on content change of provided products

### PR DESCRIPTION
**Pull request by vritant**
1514484: entitlements are regenerated on content change of provided products
* a previous commit had changed subscription resolution to no longer resolve product or content,
  this change undoes that so we can test this.
* added the spec test to test that refresh after content and product change does indeed regenerate ent

### CODE REVIEW

**Comment by Lorquas** [comment]
retest this please

File: server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java [hunk 1]
```diff
@@ -555,6 +555,21 @@ private void syncBranding(Pool pool, Pool existingPool) {
         return incomingProvided;
     }
 
+    private boolean changedProductsInSet(Set<Product> products,
+        Map<String, Product> changedProducts) {
+
+        if (products != null && changedProducts != null) {
+            for (Product product: products) {
```
Comment (Ceiu) [inline server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java]: Checkstyle failure here: need a space between product and the colon. Should also pull the method declaration into one line as well.

File: server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java [hunk 2]
```diff
@@ -555,6 +555,21 @@ private void syncBranding(Pool pool, Pool existingPool) {
...
+                if (product != null && product.getId() != null &&
```
Comment (Ceiu) [inline server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java]: You can skip the null check on the ID here. If it's null, it won't be found in the products list, nor will it cause an error.

File: server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java [hunk 3]
```diff
@@ -569,8 +584,12 @@ private boolean checkForChangedProducts(Product incomingProduct, Set<Product> in
             !currentProvided.equals(incomingProvided);
 
         // Check if the existing product is in the set of changed products
-        if (!productsChanged && changedProducts != null && pid != null) {
-            productsChanged = (changedProducts.get(pid) != null);
+        if (!productsChanged && changedProducts != null) {
+            if (pid != null) {
+                productsChanged = (changedProducts.get(pid) != null);
+            }
+            productsChanged = productsChanged ||
```
Comment (Ceiu) [inline server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java]: We can condense this whole block down to a (wrapped?) one-liner:

```
productsChanged |= (pid != null && changedProducts.get(pid) != null) || changedProductsInSet(...);
```

If you opt not to change it, then you need to fix the bug on line 589 which could revert the previous state of productsChanged if the declaration is true on 581 and the pid check is false on 589.

File: server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java [hunk 4]
```diff
@@ -607,7 +626,8 @@ private boolean checkForChangedDerivedProducts(Pool pool, Pool existingPool,
             }
         }
 
-        productsChanged = productsChanged || !currentProvided.equals(incomingProvided);
+        productsChanged = productsChanged || !currentProvided.equals(incomingProvided) ||
```
Comment (Ceiu) [inline server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java]: Not required, but you can use |= here.

File: server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java (hunk 3 above)
Comment (vritant) [reply server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java]: are you sure there is a bug?

File: server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java (hunk 1 above)
Comment (vritant) [reply server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java]: looks like our checkstyle is fine with these changes. I changed it anyway

File: server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java (hunk 2 above)
Comment (vritant) [reply server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java]: im only 50% sure on this, but during debugging I *thought* I saw some null objects. I am reluctant to try and prove it, may I ask what is the harm in an additional check or two even if that is not the case?

**Comment by vritant** [comment]
@Ceiu changes done

**Comment by Ceiu** [comment]
ACK

**Pull request merged**